src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -865,98 +1121,1221 @@ grpc_subchannel_call *grpc_client_channel_get_subchannel_call(
+    return false;
+  }
+  if (pending->batch->send_trailing_metadata &&
+      !retry_state->completed_send_trailing_metadata) {
+    return false;
+  }
+  if (pending->batch->recv_initial_metadata &&
+      !retry_state->completed_recv_initial_metadata) {
+    return false;
   }
+  if (pending->batch->recv_message &&
+      retry_state->completed_recv_message_count <
+          retry_state->started_recv_message_count) {
+    return false;
+  }
+  if (pending->batch->recv_trailing_metadata &&
+      !retry_state->completed_recv_trailing_metadata) {
+    return false;
+  }
+  return true;
 }
 
-// This is called via the call combiner, so access to calld is synchronized.
-static void waiting_for_pick_batches_resume(grpc_exec_ctx *exec_ctx,
-                                            grpc_call_element *elem) {
+// Callback used to intercept on_complete from subchannel calls.
+// Called only when retries are enabled.
+static void on_complete(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error) {
+  subchannel_batch_data *batch_data = (subchannel_batch_data *)arg;
+  grpc_call_element *elem = batch_data->elem;
   channel_data *chand = (channel_data *)elem->channel_data;
   call_data *calld = (call_data *)elem->call_data;
   if (GRPC_TRACER_ON(grpc_client_channel_trace)) {
-    gpr_log(GPR_DEBUG, "chand=%p calld=%p: sending %" PRIdPTR
-                       " pending batches to subchannel_call=%p",
-            chand, calld, calld->waiting_for_pick_batches_count,
-            calld->subchannel_call);
-  }
-  for (size_t i = 0; i < calld->waiting_for_pick_batches_count; ++i) {
-    GRPC_CLOSURE_INIT(&calld->handle_pending_batch_in_call_combiner[i],
-                      run_pending_batch_in_call_combiner, calld,
+    char *op_str = grpc_transport_stream_op_batch_string(&batch_data->batch);
+    gpr_log(GPR_DEBUG, "chand=%p calld=%p: got on_complete, error=%s, batch=%s",
+            chand, calld, grpc_error_string(error), op_str);
+    gpr_free(op_str);
+  }
+  // Update bookkeeping in retry_state.
+  subchannel_call_retry_state *retry_state = (subchannel_call_retry_state *)
+      grpc_connected_subchannel_call_get_parent_data(
+          batch_data->subchannel_call);
+  if (batch_data->batch.send_initial_metadata) {
+    retry_state->completed_send_initial_metadata = true;
+  }
+  if (batch_data->batch.send_message) {
+    ++retry_state->completed_send_message_count;
+  }
+  if (batch_data->batch.send_trailing_metadata) {
+    retry_state->completed_send_trailing_metadata = true;
+  }
+  if (batch_data->batch.recv_initial_metadata) {
+    retry_state->completed_recv_initial_metadata = true;
+  }
+  if (batch_data->batch.recv_message) {
+    ++retry_state->completed_recv_message_count;
+  }
+  if (batch_data->batch.recv_trailing_metadata) {
+    retry_state->completed_recv_trailing_metadata = true;
+  }
+  // Check if the call is finished, and if so, get its status.
+  // The call is finished if either (a) this callback was invoked with
```

Comment:
@ctiller, a question here: 

Currently, the retry code is just looking at the first error we see, whether it's a callback being invoked with `error != GRPC_ERROR_NONE` or receiving a non-OK status from recv_trailing_metadata.  However, given that those two things can happen at the same time, I am wondering if we actually need to use a heuristic similar to what surface/call.cc uses to determine the call's status based on multiple inputs.

If we do need to do something like this, then we would presumably want to wait until we get back the result from recv_trailing_metadata before we decide on the final status.  However, if the application has not yet started a recv_trailing_metadata op (e.g., if it doesn't try starting that op until recv_message returns a NULL byte buffer), then we will never see the status.  But perhaps we could work around this by having the client_channel code preemptively start a batch on the subchannel call containing only recv_trailing_metadata, so that we always get status as soon as possible, even if the application hasn't yet asked for it.

Thoughts...?